File: packages/maencof/skills/maencof-craft-skill/scripts/structure_generator.py
#!/usr/bin/env python3
"""
Structure Generator - Creates skill directory structure based on complexity

Usage:
    structure_generator.py --name <skill-name> --complexity <simple|medium|complex> --path <output-path>

Example:
    structure_generator.py --name api-client --complexity medium --path ./skills
"""

import argparse
import json
from pathlib import Path
from typing import Dict


SKILL_MD_TEMPLATES = {
    "simple": """---
name: {name}
description: [TODO: Complete description with specific trigger scenarios]
version: 1.0.0
complexity: simple
created: {date}
---

# {title}

## Quick Start
[2-3 sentence overview]

## When to Use This Skill
[Specific trigger scenarios - bulleted list]

## Workflow
[Simple 2-3 step process]

## Resources
### scripts/
[Brief description of available scripts]

## Quick Reference
[Common operations cheat sheet]
""",

    "medium": """---
name: {name}
description: [TODO: Complete description with specific trigger scenarios]
version: 1.0.0
complexity: medium
created: {date}
---

# {title}

## Quick Start
[2-3 sentence overview]

## When to Use This Skill
[Specific trigger scenarios - bulleted list]

## Core Workflow
[High-level 4-6 step process with references to reference.md]

## Resources
### reference.md
Detailed workflows, algorithms, and implementation guidance.

### examples.md
Real-world usage examples and patterns.

### scripts/
[Brief description of available automation tools]

## Quick Reference
[Common operations cheat sheet]
""",

    "complex": """---
name: {name}
description: [TODO: Complete description with specific trigger scenarios]
version: 1.0.0
complexity: complex
created: {date}
---

# {title}

## Quick Start
[2-3 sentence overview - keep brief, details in reference.md]

## When to Use This Skill
[Specific trigger scenarios - bulleted list]

## Core Workflow
[High-level overview - details in reference.md]

## Resources
### reference.md
Comprehensive workflows, algorithms, and mode-specific processes.

### examples.md
5-10 real-world examples covering all complexity levels.

### knowledge/
Deep-dive topics: skill anatomy, MCP integration, resource design, quality standards.

### scripts/
[Brief index of automation tools - details in reference.md]

### docs/
Migration guides, troubleshooting, advanced patterns.

## Quick Reference
[Essential operations only - full reference in reference.md]
"""
}


def generate_structure(name: str, complexity: str, path: Path) -> Dict:
    """
    Generate skill directory structure based on complexity.

    Returns:
        {
            "skill_path": Path,
            "files_created": List[str],
            "next_steps": List[str]
        }
    """
    from datetime import date

    skill_path = path / name

    if skill_path.exists():
        raise ValueError(f"Skill directory already exists: {skill_path}")

    skill_path.mkdir(parents=True, exist_ok=False)

    files_created = []

    # Create SKILL.md
    title = " ".join(word.capitalize() for word in name.split("-"))
    today = date.today().isoformat()
    skill_md_content = SKILL_MD_TEMPLATES[complexity].format(
        name=name,
        title=title,
        date=today
    )
    (skill_path / "SKILL.md").write_text(skill_md_content)
    files_created.append("SKILL.md")

    # Create scripts/
    scripts_dir = skill_path / "scripts"
    scripts_dir.mkdir()
    (scripts_dir / "example.py").write_text(f"#!/usr/bin/env python3\n# Example script for {name}\n")
    files_created.append("scripts/example.py")

    # Medium complexity additions
    if complexity in ["medium", "complex"]:
        (skill_path / "reference.md").write_text(f"# Detailed Reference for {title}\n\n[TODO: Add detailed workflows]\n")
        files_created.append("reference.md")

        (skill_path / "examples.md").write_text(f"# {title} Examples\n\n[TODO: Add 5-10 examples]\n")
        files_created.append("examples.md")

    # Complex additions
    if complexity == "complex":
        knowledge_dir = skill_path / "knowledge"
        knowledge_dir.mkdir()
        for topic in ["skill-anatomy.md", "progressive-disclosure.md", "mcp-integration.md",
                      "bundled-resources.md", "quality-standards.md"]:
            (knowledge_dir / topic).write_text(f"# {topic.replace('.md', '').replace('-', ' ').title()}\n\n[TODO: Add content]\n")
            files_created.append(f"knowledge/{topic}")

        docs_dir = skill_path / "docs"
        docs_dir.mkdir()
        for doc in ["migration-guide.md", "mode-comparison.md", "complexity-tuning.md",
                    "troubleshooting.md", "advanced-patterns.md"]:
            (docs_dir / doc).write_text(f"# {doc.replace('.md', '').replace('-', ' ').title()}\n\n[TODO: Add content]\n")
            files_created.append(f"docs/{doc}")

    next_steps = [
        "Edit SKILL.md to complete TODO items",
        "Update description in YAML frontmatter",
        "Implement scripts in scripts/ directory",
    ]

    if complexity in ["medium", "complex"]:
        next_steps.extend([
            "Fill in reference.md with detailed workflows",
            "Add 5-10 examples to examples.md",
        ])

    if complexity == "complex":
        next_steps.extend([
            "Complete knowledge/ deep-dive topics",
            "Write docs/ guides (migration, troubleshooting, etc.)",
        ])

    return {
        "skill_path": str(skill_path),
        "files_created": files_created,
        "next_steps": next_steps
    }


def main():
    parser = argparse.ArgumentParser(description="Generate skill directory structure")
    parser.add_argument("--name", required=True, help="Skill name (hyphen-case)")
    parser.add_argument("--complexity", required=True, choices=["simple", "medium", "complex"])
    parser.add_argument("--path", required=True, help="Output directory path")
    args = parser.parse_args()

    path = Path(args.path)
    result = generate_structure(args.name, args.complexity, path)

    print(json.dumps(result, indent=2))


if __name__ == "__main__":
    main()

File: packages/maencof/skills/maencof-craft-skill/scripts/structure_generator.mjs
#!/usr/bin/env node
/**
 * Structure Generator - Creates skill directory structure based on complexity
 *
 * Usage:
 *     node structure_generator.mjs --name <skill-name> --complexity <simple|medium|complex> --path <output-path>
 *
 * Example:
 *     node structure_generator.mjs --name api-client --complexity medium --path ./skills
 */

import { parseArgs } from 'node:util';
import { existsSync, mkdirSync, writeFileSync } from 'node:fs';
import path from 'node:path';

const SKILL_MD_TEMPLATES = {
  simple: `---
name: {name}
description: [TODO: Complete description with specific trigger scenarios]
---

# {title}

## Quick Start
[2-3 sentence overview]

## When to Use This Skill
[Specific trigger scenarios - bulleted list]

## Workflow
[Simple 2-3 step process]

## Resources
### scripts/
[Brief description of available scripts]

## Quick Reference
[Common operations cheat sheet]
`,

  medium: `---
name: {name}
description: [TODO: Complete description with specific trigger scenarios]
---

# {title}

## Quick Start
[2-3 sentence overview]

## When to Use This Skill
[Specific trigger scenarios - bulleted list]

## Core Workflow
[High-level 4-6 step process with references to reference.md]

## Resources
### reference.md
Detailed workflows, algorithms, and implementation guidance.

### examples.md
Real-world usage examples and patterns.

### scripts/
[Brief description of available automation tools]

## Quick Reference
[Common operations cheat sheet]
`,

  complex: `---
name: {name}
description: [TODO: Complete description with specific trigger scenarios]
---

# {title}

## Quick Start
[2-3 sentence overview - keep brief, details in reference.md]

## When to Use This Skill
[Specific trigger scenarios - bulleted list]

## Core Workflow
[High-level overview - details in reference.md]

## Resources
### reference.md
Comprehensive workflows, algorithms, and mode-specific processes.

### examples.md
5-10 real-world examples covering all complexity levels.

### knowledge/
Deep-dive topics: skill anatomy, MCP integration, resource design, quality standards.

### scripts/
[Brief index of automation tools - details in reference.md]

### docs/
Migration guides, troubleshooting, advanced patterns.

## Quick Reference
[Essential operations only - full reference in reference.md]
`,
};

function generateStructure(name, complexity, outputPath) {
  const skillPath = path.join(outputPath, name);

  if (existsSync(skillPath)) {
    throw new Error(`Skill directory already exists: ${skillPath}`);
  }

  mkdirSync(skillPath, { recursive: true });

  const filesCreated = [];

  // Create SKILL.md
  const title = name
    .split('-')
    .map((w) => w.charAt(0).toUpperCase() + w.slice(1))
    .join(' ');
  const skillMdContent = SKILL_MD_TEMPLATES[complexity]
    .replace(/{name}/g, name)
    .replace(/{title}/g, title);
  writeFileSync(path.join(skillPath, 'SKILL.md'), skillMdContent);
  filesCreated.push('SKILL.md');

  // Create scripts/
  const scriptsDir = path.join(skillPath, 'scripts');
  mkdirSync(scriptsDir);
  writeFileSync(
    path.join(scriptsDir, 'example.mjs'),
    `#!/usr/bin/env node\n// Example script for ${name}\n`
  );
  filesCreated.push('scripts/example.mjs');

  // Medium complexity additions
  if (complexity === 'medium' || complexity === 'complex') {
    writeFileSync(
      path.join(skillPath, 'reference.md'),
      `# Detailed Reference for ${title}\n\n[TODO: Add detailed workflows]\n`
    );
    filesCreated.push('reference.md');

    writeFileSync(
      path.join(skillPath, 'examples.md'),
      `# ${title} Examples\n\n[TODO: Add 5-10 examples]\n`
    );
    filesCreated.push('examples.md');
  }

  // Complex additions
  if (complexity === 'complex') {
    const knowledgeDir = path.join(skillPath, 'knowledge');
    mkdirSync(knowledgeDir);
    for (const topic of [
      'skill-anatomy.md',
      'progressive-disclosure.md',
      'mcp-integration.md',
      'bundled-resources.md',
      'quality-standards.md',
    ]) {
      const topicTitle = topic
        .replace('.md', '')
        .replace(/-/g, ' ')
        .replace(/\b\w/g, (c) => c.toUpperCase());
      writeFileSync(
        path.join(knowledgeDir, topic),
        `# ${topicTitle}\n\n[TODO: Add content]\n`
      );
      filesCreated.push(`knowledge/${topic}`);
    }

    const docsDir = path.join(skillPath, 'docs');
    mkdirSync(docsDir);
    for (const doc of [
      'migration-guide.md',
      'mode-comparison.md',
      'complexity-tuning.md',
      'troubleshooting.md',
      'advanced-patterns.md',
    ]) {
      const docTitle = doc
        .replace('.md', '')
        .replace(/-/g, ' ')
        .replace(/\b\w/g, (c) => c.toUpperCase());
      writeFileSync(
        path.join(docsDir, doc),
        `# ${docTitle}\n\n[TODO: Add content]\n`
      );
      filesCreated.push(`docs/${doc}`);
    }
  }

  const nextSteps = [
    'Edit SKILL.md to complete TODO items',
    'Update description in YAML frontmatter',
    'Implement scripts in scripts/ directory',
  ];

  if (complexity === 'medium' || complexity === 'complex') {
    nextSteps.push(
      'Fill in reference.md with detailed workflows',
      'Add 5-10 examples to examples.md'
    );
  }

  if (complexity === 'complex') {
    nextSteps.push(
      'Complete knowledge/ deep-dive topics',
      'Write docs/ guides (migration, troubleshooting, etc.)'
    );
  }

  return {
    skill_path: skillPath,
    files_created: filesCreated,
    next_steps: nextSteps,
  };
}

const { values } = parseArgs({
  options: {
    name: { type: 'string' },
    complexity: { type: 'string' },
    path: { type: 'string' },
  },
  strict: true,
});

if (!values.name || !values.complexity || !values.path) {
  console.error('Usage: node structure_generator.mjs --name <name> --complexity <simple|medium|complex> --path <output-path>');
  process.exit(1);
}

if (!['simple', 'medium', 'complex'].includes(values.complexity)) {
  console.error(`Invalid complexity: ${values.complexity}. Must be simple, medium, or complex.`);
  process.exit(1);
}

const result = generateStructure(values.name, values.complexity, path.resolve(values.path));
console.log(JSON.stringify(result, null, 2));
